### DISCUSSION TITLE
Logtalk for multi agent simulation?

### DISCUSSION THREAD

**Discussion by EllenOrange** [LogtalkDotOrg/logtalk3#179] [Q&A] [answered]
Hi folks,

I'm building a multi-agent simulation for a game I'm developing and I'm curious if y'all think LogTalk can work as the simulation language.

The key requirements for the system are:
1) Specify ontologies that define a hierarchy of classes with properties with types and default values, and a hierarchy of actions over those classes.
2) Defeasible inference of property values, with instance properties defeating class properties, and child properties defeating parent properties.
3) Facts defined over the ontology, organized by objects.
4) Multiple independent simulations with different ontologies and facts
5) Update facts AND ontology for any simulation, at runtime

This will then be used to create an simulation of the physical world, and separate simulations of the mind of each of the characters. Minds can update based on observations of the physical world. 

I've prototyped the above in SWI-Prolog and achieved 3) and 4) via modules and meta predicates, but the code needed to define even a simple ontology will make for a poor authoring experience. I can clean this up by implementing a DSL via a Prolog DCG, but I see no good way of providing editor and debugging support for the DSL. 

This is where I'm hoping Logtalk might come in. 1), 2), and 3) seem to be exactly what Logtalk is made for, and Logtalk already has editor and debugger support. It seems like I could accomplish 4) by making a Logtalk object for each simulator and including the ontology and facts into that simulation. 

It seems like 5) is the main issue, especially when it comes to updating the ontology at runtime; once a category is defined by :- category/1 ... :end_category or create_category/4, there doesn't seem to be any way of incrementally updating the category. One could abolish it and create a new definition with the desired changes, but it's unclear if existing instances of that category would continue to work sensibly afterward. 

Thoughts?

I'm especially interested in options for resolving 5) and whether I'm correct that 2) is straight forward.

**Comment by pmoura** [comment, marked as answer]
> Hi folks,
> 
> I'm building a multi-agent simulation for a game I'm developing and I'm curious if y'all think LogTalk can work as the simulation language.

Intersting. Btw, one of the first commercial uses of Logtalk (+20 years ago) was a game based on historical events.

> The key requirements for the system are:
> 
> 1. Specify a ontologies that define a hierarchy of classes with properties with types and default values, and a hierarchy of actions over those classes.
> 2. Defeasible inference of property values, with instance properties defeating class properties, and child properties defeating parent properties.
> 3. Facts defined over the ontology, organized by objects.

1, 2, and 3 should be straight-forward. You may want to look into using _prototypes_ instead of _classes_ and _instances_ for simplicity, however. Also, using _categories_ to simplify hierarchies. Related:

https://logtalk.org/2020/02/24/a-category-at-the-top.html

Regarding defeasible inference, by default, descendant properties override inherited properties. See the Handbook section on inheritance:

https://logtalk.org/manuals/userman/inheritance.html

> 4. Multiple independent simulations with different ontologies and facts
> 5. Update facts AND ontology for any simulation, at runtime

The "many worlds" design pattern is likely a good fit here. For a very simple example, see:

https://logtalk.org/2019/11/13/many-worlds-design-pattern.html

A more serious example can be found in the Metagol port (an ILP system):

https://github.com/LogtalkDotOrg/logtalk3/tree/master/ports/metagol

In this case, the common knowledge is represented in the `metagol` object and examples are the different worlds, all implementing a common protocol. Each example/world contains its own facts and rules. Updates can be easily accomplished by declaring and defining object dynamic predicates.

Is your current prototype available publicly?

**Reply by EllenOrange** [reply]
Thanks for the speedy reply and the links! I'll check them out right away; I'm excited that Logtalk might be what I'm looking for.

> Btw, one of the first commercial uses of Logtalk (+20 years ago) was a game based on historical events.

Nice! What game is that?

> Is your current prototype available publicly?

Alas, no. It's a messy prototype intended to help me discover the real requirements for the system and to learn Prolog along the way. Mission accomplished but there is no way it's seeing the light of day. :)

**Comment by pmoura** [comment]
> > Btw, one of the first commercial uses of Logtalk (+20 years ago) was a game based on historical events.
> 
> Nice! What game is that?

The historical context was one of the French invasions of Portugal and the uprise against the invaders in the village of Soure in the 1800s. Logtalk was used to represent  the characters in the game and their behavior: soldiers, priests, ... The game GUI was written in VisualBasic with Logtalk running on LPA Prolog. Fun times. I still have some screenshots somewhere in one of my old computers.

**Reply by EllenOrange** [reply]
Sweet! 

It's smart that you kept screenshots; there are definitely some games prototypes I made back at Maxis before The Sims launched, that I wish I'd at least kept images of.
